bee-network/bee-autopeering/src/discovery/manager.rs

```diff
@@ -0,0 +1,1000 @@
+    peer_addr: SocketAddr,
+    peer_id: &PeerId,
+    verif_req: &VerificationRequest,
+    msg_bytes: &[u8],
+    server_tx: &ServerTx,
+    local: &Local,
+) {
+    log::trace!("Sending verification response to: {}/{}", peer_id, peer_addr);
+
+    let request_hash = msg_hash(MessageType::VerificationRequest, msg_bytes).to_vec();
+
+    let verif_res = VerificationResponse::new(request_hash, local.services(), peer_addr.ip());
+
+    let msg_bytes = verif_res
+        .to_protobuf()
+        .expect("error encoding verification response")
+        .to_vec();
+
+    // hive.go:
+    // ```go
+    // // the destination address uses the source IP address of the packet plus the src_port from the message
+    // dstAddr := &net.UDPAddr{
+    // 	IP:   fromAddr.IP,
+    // 	Port: int(m.SrcPort),
+    // }
+    // ```
+    server_tx
+        .send(OutgoingPacket {
+            msg_type: MessageType::VerificationResponse,
+            msg_bytes,
+            peer_addr: SocketAddr::new(peer_addr.ip(), verif_req.source_addr().port()),
+        })
+        .expect("error sending verification response to server");
+}
+
+/// Initiates a discovery request to a peer by fetching its endpoint data from the peer store and waiting
+/// for the peer's response, which must arrive in time.
+///
+/// Returns `Some(Vec<Peer>)` of discovered peers, if the request was successful, otherwise `None`.
+pub(crate) async fn begin_discovery(
+    peer_id: &PeerId,
+    active_peers: &ActivePeersList,
+    request_mngr: &RequestManager,
+    server_tx: &ServerTx,
+) -> Option<Vec<Peer>> {
+    let (response_tx, response_rx) = request::response_chan();
+
+    send_discovery_request_to_peer(peer_id, active_peers, request_mngr, server_tx, Some(response_tx));
+
+    match tokio::time::timeout(RESPONSE_TIMEOUT, response_rx).await {
+        Ok(Ok(bytes)) => {
+            let disc_res = DiscoveryResponse::from_protobuf(&bytes).expect("error decoding discovery response");
+            Some(disc_res.into_peers())
+        }
+        Ok(Err(e)) => {
+            // This shouldn't happen under normal circumstances.
+            log::debug!("Discovery response error: {}", e);
+            Some(Vec::new())
+        }
+        Err(e) => {
+            log::debug!("Discovery response timeout: {}", e);
+
+            // The response didn't arrive in time => remove the request.
+            let _ = request_mngr.write().pull::<DiscoveryRequest>(peer_id);
+
+            None
+        }
+    }
+}
+
+/// Sends a discovery request to a peer by fetching its endpoint data from the peer store.
+///
+/// The function is non-blocking.
+pub(crate) fn send_discovery_request_to_peer(
+    peer_id: &PeerId,
+    active_peers: &ActivePeersList,
+    request_mngr: &RequestManager,
+    server_tx: &ServerTx,
+    response_tx: Option<ResponseTx>,
+) {
+    let guard = active_peers.read();
+    let active_peer = guard.find(peer_id).expect("peer not in active peers list");
```

Comment:
If you're not keeping guards longer I'd just chain it to be 100% sure it's dropped, sometimes it behaves weirdly

Reply:
But `guard`'s lifetime needs to exceed `active_peer`s lifetime, because `find` returns `Option<&P`.

Reply:
Alright...I know what you mean and wanted, and I fully agree. Done.